2018/004-image-update-caasp-v4.md

```diff
@@ -0,0 +1,660 @@
+> change once this change is implemented.
+
+We are proposing the following upgrade policies (please look at the previous
+section for a detailed explanation of the different cases):
+
+  * Case #1 - core application upgrade: we want to be fully in charge of that.
+    New salt states and kubernetes manifests must be available on the admin node
+    before we can use these images. They cannot be automated.
+  * Case #2 - same core application upgrade, only fixes to the image: in some
+    corner cases we could still need changes to the salt states and/or the
+    kubernetes manifests. Hence, they cannot be automated.
+  * Case #3 - maintenance update: no changes needed neither to salt states nor
+    to kubernetes manifests. These can be automated.
+  * Case #4 - update of core application via SLE maintainer: we think the best
+    solution to this scenario is to have the maintenance QA block this automatic
+    rebuild of the image from being published. They should notify the development
+    team about this core application upgrade so that everything can fall back to
+    case #1.
+
+It's important to point out that we are not looking for the perfect solution,
+we are looking for something that allows us to release CaaSP 4.0 on time,
+without introducing technical debt, without making the whole upgrade story worse
+than the one of v3 and without having to write that much code.
+
+V3 already has some limits, we are fine keeping them around for 4.0:
+
+  * Nodes have to be rebooted to get latest images.
+  * Admin nodes images are always using the latest version available on the
+    node.
+  * Nodes are rebooted even when not actually needed: images are delivered via
+    RPMs, when a new velum image is published all the nodes of the cluster are
+    going to be rebooted even if the only one that actually needs this package
+    is the admin node.
+  * Kubernetes add-ons (dex, tiller, kubedns) are currently referenced by their
+    short tag.
+  * Kubernetes daemon-sets (flannel) are referenced by their short tag.
+  * Static pods on the Kubernetes nodes (haproxy) are referenced by their
+    short tag.
+  * The short tag consists only of the version of the core application shipped
+    inside of the image. If the image is rebuilt due to a bug or a security fix
+    the short tag won't change. The usage of short tags can lead customers to
+    run unpatched images or bugged ones without being aware of that.
+
+The proposed solution for 4.0 is going to keep some of these limitations (reboots)
+but is going to improve the security aspects.
+
+## Detailed RFC
+
+> In this section of the document the target audience is the dev team. Upon
+> reading this section each engineer should have a rather clear picture of what
+> needs to be done in order to implement the described feature. 
+
+### Changes for image building
+
+Images are going to have 3 tags:
+
+  * **short:** this is the core application version. For example, given `2.6.2`
```

Comment:
This means that if there is a minor version update (e.g. 2.6.3 contains security fixes) the referenced tag has to change as well?